### ISSUE TITLE
Passing trait objects to C++

### ISSUE THREAD

**Issue by Timmmm** [dtolnay/cxx#665]
Suppose I have some trait:

```rust
trait MyData {
  ...
}
```

And I want to pass a trait object to C++:

```rust
#[cxx::bridge]
mod ffi {
  extern "Rust" {
    type MyData;
    pub fn read_data() -> Result<Box<dyn MyData>>;
  }
}
```

It says

    error[E0277]: the size for values of type `(dyn MyData + 'static)` cannot be known at compilation time

This appears to be consistent with [this comment](https://cxx.rs/binding/box.html#restrictions):

> If T is an opaque Rust type, the Rust type is required to be Sized i.e. size known at compile time. In the future we may introduce support for dynamically sized opaque Rust types.

Is there a workaround?

**Comment by dtolnay** [comment]
You can use Box\<dyn MyData\> (fat pointer) as the opaque type.

```rust
trait MyData {...}
type DynMyData = Box<dyn MyData>;

#[cxx::bridge]
mod ffi {
    extern "Rust" {
        type DynMyData;
        fn read_data() -> Result<Box<DynMyData>>;
    }
}
```

**Comment by Timmmm** [comment]
Ah that worked, thank you! I was pretty close actually - I tried `Box<Box<dyn MyData>>` but it didn't work unfortunately. Bit awkward having to double-box and if I understand things correctly it shouldn't be strictly necessary since the first `Box` is already sized. Is this something that could be fixed?

**Comment by dtolnay** [comment]
> Is this something that could be fixed?

Yes, that is what the documentation you quoted means by: *"In the future we may introduce support for dynamically sized opaque Rust types."* It just needs a PR.

**Comment by Timmmm** [comment]
D'oh I can't read. Sorry!

**Comment by dtolnay** [comment]
> Bit awkward having to double-box and if I understand things correctly it shouldn't be strictly necessary since the first `Box` is already sized.

I opened #672 to show how to do it without a double layer of `Box`. Without direct support from the library it's probably more work than double box, but could be worth it if there are lots more signatures involving the trait object than there are trait methods.

**Comment by Timmmm** [comment]
Thanks, that's really helpful!
